### ISSUE TITLE
Bug: GitHub/GraphQL connector does not respect LIMIT

### ISSUE THREAD

**Issue by peasee** [spiceai/spiceai#2850]
## Describe the bug

When running a query with a LIMIT in the GitHub/GraphQL connector, it does not respect it and instead returns all results:

```
sql> select count(*) from spiceai.issues
+----------+
| count(*) |
+----------+
| 931      |
+----------+

Time: 17.888681966 seconds. 1 rows.
sql> select count(*) from spiceai.issues limit 200
+----------+
| count(*) |
+----------+
| 931      |
+----------+

Time: 17.182362023 seconds. 1 rows.
```

## To Reproduce

1. Create a spicepod using the GitHub issues connector
2. Run a query with a limit
4. Observe unlimited rows

## Expected behavior

The rows should be limited as expected

### Spicepod

```yaml
version: v1beta1
kind: Spicepod
name: spicepod-graphql-spicehq-decisions

datasets:
  - from: github:github.com/spiceai/spiceai/issues
    name: spiceai.issues
    params:
      github_token: ${secrets:GITHUB_TOKEN}
```

**Have you tried this on the latest `trunk` branch?**

- [x] Yes
- [ ] No

**If possible, run `spiced` with DEBUG log level**
By setting the environment variable `SPICED_LOG="task_history=DEBUG,spiced=DEBUG,runtime=DEBUG,secrets=DEBUG,data_components=DEBUG,cache=DEBUG,extensions=DEBUG,spice_cloud=DEBUG"`

**Comment by Jeadie** [comment]
Using 
```yaml
datasets: 
  - name: me
    from: graphql:https://api.github.com/graphql
    params:
      graphql_auth_token: ${env:GITHUB_TOKEN}
      unnest_depth: 1
      graphql_query: |
        {
          repository(owner: "jeadie", name: "my-first-app") {
            defaultBranchRef {
                target {
                    ... on Commit {
                        history(first: 100) {
                            pageInfo {
                                hasNextPage
                                endCursor
                            }
                            nodes {
                                message
                                message_head_line: messageHeadline
                                message_body: messageBody
                                sha: oid
                                additions
                                deletions
                                id
                                committed_date: committedDate
                                authorName: author {
                                    author_name: name
                                }
                                authorEmail: author {
                                    author_email: email
                                }
                            }
                        }
                    }
                }
            }
          }
        }
```

Running
```shell
>> curlie -XPOST "http://localhost:8090/v1/sql" --data "select * from me limit 150" | jq '.[].sha' | wc -l
     150
```

**Comment by Jeadie** [comment]
Using spicepod mentioned above
```shell
sql> select count(*) from spiceai.issues limit 200

+----------+
| count(*) |
+----------+
| 934      |
+----------+

Time: 19.26027625 seconds. 1 rows.
```

I think this might be an issue with `count(*)`. For example
```shell
>>> curlie -XPOST "http://localhost:8090/v1/sql" --data " select id from spiceai.issues limit 100" \
    | jq '.[].id' \
    | wc -l

     100
```

**Comment by Jeadie** [comment]
```shell
sql> explain select id from spiceai.issues limit 100

+---------------+----------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                               |
+---------------+----------------------------------------------------------------------------------------------------+
| logical_plan  | Limit: skip=0, fetch=100                                                                           |
|               |   BytesProcessedNode                                                                               |
|               |     TableScan: spiceai.issues projection=[id]                                                      |
| physical_plan | GlobalLimitExec: skip=0, fetch=100                                                                 |
|               |   CoalescePartitionsExec                                                                           |
|               |     BytesProcessedExec                                                                             |
|               |       MemoryExec: partitions=10, partition_sizes=[100, 100, 100, 100, 100, 100, 100, 100, 100, 34] |
|               |                                                                                                    |
+---------------+----------------------------------------------------------------------------------------------------+
```

```shell
sql> explain select count(*) from spiceai.issues limit 200

+---------------+--------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                   |
+---------------+--------------------------------------------------------------------------------------------------------+
| logical_plan  | Limit: skip=0, fetch=200                                                                               |
|               |   Aggregate: groupBy=[[]], aggr=[[count(Int64(1)) AS count(*)]]                                        |
|               |     BytesProcessedNode                                                                                 |
|               |       TableScan: spiceai.issues projection=[]                                                          |
| physical_plan | GlobalLimitExec: skip=0, fetch=200                                                                     |
|               |   AggregateExec: mode=Final, gby=[], aggr=[count(*)]                                                   |
|               |     CoalescePartitionsExec                                                                             |
|               |       AggregateExec: mode=Partial, gby=[], aggr=[count(*)]                                             |
|               |         BytesProcessedExec                                                                             |
|               |           MemoryExec: partitions=10, partition_sizes=[100, 100, 100, 100, 100, 100, 100, 100, 100, 34] |
|               |                                                                                                        |
+---------------+--------------------------------------------------------------------------------------------------------+

Time: 16.757959208 seconds. 2 rows.
```

**Comment by Jeadie** [comment]
@peasee  confirmed that it is an issue with `GraphQLTableProvider::scan(` always receiving `limit: Option<usize> == None` 

Had a similar issue, fixed in #2564, that was caused by `BytesProcessedNode`.

**Comment by Jeadie** [comment]
Confirmed `BytesProcessedNode` is not the issue.
#### How
 - Removed [BytesProcessedAnalyzerRule](https://github.com/spiceai/spiceai/blob/7cf483ddbed7a27e7c0e6651141ac8e93493e383/crates/runtime/src/datafusion.rs#L283) from Datafusion context.

#### Result
```shell
sql> select count(*) from spiceai.issues limit 200

+----------+
| count(*) |
+----------+
| 934      |
+----------+

Time: 18.160008667 seconds. 1 rows.

sql> explain select count(*) from spiceai.issues limit 200
+---------------+-----------------------------------------------------------------+
| plan_type     | plan                                                            |
+---------------+-----------------------------------------------------------------+
| logical_plan  | Limit: skip=0, fetch=200                                        |
|               |   Aggregate: groupBy=[[]], aggr=[[count(Int64(1)) AS count(*)]] |
|               |     TableScan: spiceai.issues projection=[]                     |
| physical_plan | ProjectionExec: expr=[934 as count(*)]                          |
|               |   GlobalLimitExec: skip=0, fetch=200                            |
|               |     PlaceholderRowExec                                          |
|               |                                                                 |
+---------------+-----------------------------------------------------------------+
```

**Comment by Jeadie** [comment]
Explain looks odd

```shell
sql> explain select count(*) from spiceai.issues limit 200
+---------------+-----------------------------------------------------------------+
| plan_type     | plan                                                            |
+---------------+-----------------------------------------------------------------+
| logical_plan  | Limit: skip=0, fetch=200                                        |
|               |   Aggregate: groupBy=[[]], aggr=[[count(Int64(1)) AS count(*)]] |
|               |     TableScan: spiceai.issues projection=[]                     | # No limit/fetch=200 at base `TableScan`.
| physical_plan | ProjectionExec: expr=[934 as count(*)]                          | # Harcoding 934 as `count(*)`?
|               |   GlobalLimitExec: skip=0, fetch=200                            | # LIMIT 200 is pushed down to this, to ensure single row from `PlaceholderRowExec` is < 200.
|               |     PlaceholderRowExec                                          | # Execution plan for empty relation
|               |                                                                 |
+---------------+-----------------------------------------------------------------+
```

I suspect internally, this
```sql
select count(*) from spiceai.issues limit 200
```
Becomes
```sql
(select count(*) from spiceai.issues) limit 200
```

**Comment by Jeadie** [comment]
Aggregate functions are causing issue with limit pushdowns (even if the above SQL mentioned in the bug is not invalid)
```shell
sql> explain select count(*) from (select * from spiceai.issues limit 200)
+---------------+------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                       |
+---------------+------------------------------------------------------------------------------------------------------------+
| logical_plan  | Aggregate: groupBy=[[]], aggr=[[count(Int64(1)) AS count(*)]]                                              |
|               |   Limit: skip=0, fetch=200                                                                                 |
|               |     BytesProcessedNode                                                                                     |
|               |       TableScan: spiceai.issues projection=[]                                                              |
| physical_plan | AggregateExec: mode=Final, gby=[], aggr=[count(*)]                                                         |
|               |   CoalescePartitionsExec                                                                                   |
|               |     AggregateExec: mode=Partial, gby=[], aggr=[count(*)]                                                   |
|               |       RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1                                |
|               |         GlobalLimitExec: skip=0, fetch=200                                                                 |
|               |           CoalescePartitionsExec                                                                           |
|               |             BytesProcessedExec                                                                             |
|               |               MemoryExec: partitions=10, partition_sizes=[100, 100, 100, 100, 100, 100, 100, 100, 100, 34] |
|               |                                                                                                            |
+---------------+------------------------------------------------------------------------------------------------------------+

Time: 17.122319125 seconds. 2 rows.
```

**Comment by Jeadie** [comment]
Like [above](https://github.com/spiceai/spiceai/issues/2850#issuecomment-2381962778), no issue without `BytesProcessedExec`
```shell
sql> explain select count(*) from (select * from spiceai.issues limit 200)
+---------------+---------------------------------------------------------------+
| plan_type     | plan                                                          |
+---------------+---------------------------------------------------------------+
| logical_plan  | Aggregate: groupBy=[[]], aggr=[[count(Int64(1)) AS count(*)]] |
|               |   Limit: skip=0, fetch=200                                    |
|               |     TableScan: spiceai.issues projection=[], fetch=200        |
| physical_plan | ProjectionExec: expr=[200 as count(*)]                        |
|               |   PlaceholderRowExec                                          |
|               |                                                               |
+---------------+---------------------------------------------------------------+

Time: 3.186298 seconds. 2 rows.
```

**Comment by Jeadie** [comment]
Hypothesis: [PushDownLimit](https://github.com/apache/datafusion/blob/f1aa27fdb292c562173560224ea8cda57d5d78e4/datafusion/optimizer/src/push_down_limit.rs#L45) can push down limit for [TableScan](https://github.com/apache/datafusion/blob/f1aa27fdb292c562173560224ea8cda57d5d78e4/datafusion/optimizer/src/push_down_limit.rs#L86), but won't push down for [BytesProcessedNodes](https://github.com/apache/datafusion/blob/f1aa27fdb292c562173560224ea8cda57d5d78e4/datafusion/optimizer/src/push_down_limit.rs#L156).

**Comment by Jeadie** [comment]
Above is correct, We should rewrite [BytesProcessedNodes](https://github.com/apache/datafusion/blob/f1aa27fdb292c562173560224ea8cda57d5d78e4/datafusion/optimizer/src/push_down_limit.rs#L156) as an optimizer rule instead and
1. Make it handle `LogicalPlan::Limit(LogicalPlan::TableScan{})`
2. Apply it after [PushDownLimit](https://github.com/apache/datafusion/blob/f1aa27fdb292c562173560224ea8cda57d5d78e4/datafusion/optimizer/src/push_down_limit.rs#L45)

**Comment by phillipleblanc** [comment]
Fixed in #2867
